File: s2/c11.cpp
#include <unistd.h>
#include <string>
#include "utils.h"
#include "aes128_ecb.h"
#include "aes128_cbc.h"
#include "modeECB.h"

using namespace std;

class allMightyOracle : public encOracle {
	int m_prev;
	int m_ecbGCorrect;
	int m_cbcGCorrect;
	int m_total;
	public:
	allMightyOracle() : m_prev(-1), m_ecbGCorrect(0), m_cbcGCorrect(0), m_total(0) {}
	virtual string encrypt(const string& input) override {
		return encryptRandomly(input);
	}
	string encryptRandomly(const string& input) {
		//	prefix
		string data = createRandomString(rand()%5 + 5);
		data += input;
		//	suffix
		data += createRandomString(rand()%5 + 5);
		//	padding
		data = pad(data, 16);
		string key = createRandomString(16);
		string iv = createRandomString(16);
		aes128_ecb ecb(key);
		aes128_cbc cbc(key);
		m_prev = rand();
		m_total++;
		if(m_prev%2 == 0) {
			return ecb.encrypt(data);
		}
		return cbc.encrypt(data, iv);
	}
	void guessECB() {
		if(m_prev != -1 && m_prev%2 == 0)
			m_ecbGCorrect++;
		m_prev = -1;
	}
	void guessCBC() {
		if(m_prev != -1 && m_prev%2 != 0)
			m_cbcGCorrect++;
		m_prev = -1;
	}
	void showStats() {
		log("Tested:\t%d\nECB:\t%d\nCBC:\t%d\n%s", m_total, m_ecbGCorrect,
			   m_cbcGCorrect, (m_total == m_cbcGCorrect + m_ecbGCorrect)?"All guessed OK":"FAILED!!!");
	}
};

int main() {
	srand(time(NULL));
	allMightyOracle oracle;
	string data(48, 'A');
	for(int i = 0; i < 256; i++) {
		if(isECB(oracle, 16, 4))
			oracle.guessECB();
		else
			oracle.guessCBC();
	}
	oracle.showStats();
	return 0;
}

File: s2/c11.py
#!/usr/bin/python3

import py_common.cryptoUtils as cu
from random import SystemRandom as SRand

def randomEncryption(plain):
    iv = cu.createRandomString(16)
    key = cu.createRandomString(16)
    #   prepend with random
    plain = cu.createRandomString(SRand().randint(5, 10)) + plain
    #   append with random
    plain += cu.createRandomString(SRand().randint(5, 10))
    #   pad
    plain = cu.pad(plain, 16)
    if SRand().randint(0,1):
        return (cu.encryptAES_ECB(plain, key),"ECB")
    else:
        return (cu.encryptAES_CBC(plain, key, iv),"CBC")

if __name__ == "__main__":
    numTests = 256
    data = b"A"*48
    ecb = 0
    cbc = 0
    for i in range(0, numTests):
        (cryptxt, truth) = randomEncryption(data)
        guess = cu.isECB(cryptxt)
        if guess == True and truth == "ECB":
            ecb+=1
        elif truth == "CBC":
            cbc += 1
    print("Tested:\t",numTests)
    print("ECB:\t",ecb)
    print("CBC:\t",cbc)
    if ecb+cbc == numTests:
        print("All guessed OK")
    else:
        print("Some guesses mismatched!!!")
